### ISSUE TITLE
Virtualization support in raspberry pi kernel

### ISSUE THREAD

**Issue by hashken** [raspberrypi/linux#377]
How much virtualization support is enabled in the kernel by default.

Specifically, I want to run mininet on raspberry pi and mininet.

Mininet's <a href=https://github.com/mininet/mininet>Github page</a> lists this

Mininet creates virtual networks using process-based virtualization and network namespaces - features that are available in recent Linux kernels.

I want to know if the raspberry pi kernel can support this by default.

**Comment by popcornmix** [comment]
```
$ zcat /proc/config.gz |grep -i VIRTUALIZATION
# CONFIG_VIRTUALIZATION is not set
```

So, I'm guessing no. My suspicion is that virtualizion support would add some overhead to the kernel (when not being used), so I'd need some evidence that this option doesn't cause an overhead (in CPU/memory) and is wanted by a number of people.

If you want to play with config options, here would be a good start:
http://elinux.org/RPi_Kernel_Compilation

**Comment by Ferroin** [comment]
CONFIG_VIRTUALIZATION is a dummy config that activates a submenu for KVM and virtio-net drivers, neither of which should be needed for this.

If unspecified Namespace support is enabled by deefault, including network namespaces.  Additionally, 'process based virtualization' is marketing gibberish for process sandboxing, which is done using standard syscalls that are always in the kernel (ie chroot).

**Comment by vk5tu** [comment]
The usual reason for running Mininet is to build a virtual network of OpenVSwitches, often controlled by an OpenFlow controller.

$ gzip -d -c /proc/config.gz | grep CONFIG_OPENVSWITCH
# CONFIG_OPENVSWITCH is not set

**Comment by hashken** [comment]
Isn't openvswitch inserted as a module when you compile its source. Are you sure that there is a separate config option for that in the Linux kernel?

**Comment by vk5tu** [comment]
> Isn't openvswitch inserted as a module when you compile its source.

http://git.openvswitch.org/cgi-bin/gitweb.cgi?p=openvswitch;a=blob_plain;f=FAQ;hb=HEAD
"Q: What virtualization platforms can use Open vSwitch?
A: ... As of Linux 3.3 it is part of the mainline kernel...."

> Are you sure that there is a separate config option for that in the Linux kernel?

$ grep Hardware /proc/cpuinfo 
Hardware    : BCM2708
$ uname -r -v
3.6.11+ #538 PREEMPT Fri Aug 30 20:42:08 BST 2013
$ gzip -d -c /proc/config.gz | grep CONFIG_OPENVSWITCH
# CONFIG_OPENVSWITCH is not set

**Comment by popcornmix** [comment]
Are we happy CONFIG_OPENVSWITCH is all that is needed for mininet?
As it's a module, then no problem in enabling it.

**Comment by vk5tu** [comment]
Checking the documentation for OpenVSwitch, it looks like it also needs VLANs, and that's configured anyway.

Checking the documentation for Mininet itself:

https://github.com/mininet/mininet/blob/master/INSTALL
"5. Installation on other Linux distributions
   In general, you must have:
- A Linux kernel compiled with network namespace support enabled
- ..."

$ gzip -d -c /proc/config.gz | grep CONFIG_NET_NS
CONFIG_NET_NS=y

So that looks good too.

In short, I'm confident but at the same time wouldn't be shocked if some minor option was missing. Is that a high enough bar? And if not then what do you need?

**Comment by popcornmix** [comment]
Sounds good enough for me. I'll add it with next update.

**Comment by vk5tu** [comment]
Thanks heaps. Give me a ping when the next update happens and I'll install mininet on the RPi and put the directions on my blog for others. The RPi should be a good platform for messing about with OpenFlow, as controller scaling problems can be made to happen using a average PC rather than needing professional test gear.

**Comment by popcornmix** [comment]
Please update and test:

```
pi@raspberrypi:~ $ uname -a
Linux raspberrypi 3.10.12+ #549 Tue Sep 24 00:41:33 BST 2013 armv6l GNU/Linux
pi@raspberrypi:~ $ zcat /proc/config.gz | grep CONFIG_OPENVSWITCH
CONFIG_OPENVSWITCH=m
```

**Comment by vk5tu** [comment]
Tested and works. Documented at [Installing OpenVSwitch and Mininet on Raspberry Pi](http://vk5tu.livejournal.com/51207.html)
